### PR TITLE
Skip `test_disk_usage` if there's no `/usr/bin/du`

### PR DESCRIPTION
**Pull request by hawkw**
The `zone_bundle::disk_usage` function in sled-agent executes `du(1)` to
measure disk usage for a path. Presently, the path to the `du(1)`
executable is hard-coded to `/usr/bin/du`. On some systems, such as
NixOS, the `du(1)` executable may not live at the path `/usr/bin/du`. We
could, potentially, change the `disk_usage` function to find a `du(1)`
binary on the PATH, or fall back to other locations...but, on Helios, it
is always at `/usr/bin/du`, so hardcoding it is a good solution for the
environment that sled-agents are actually intended to run in.

However, hardcoding the path means that this test will always fail on
systems where there is no `/usr/bin/du`. Therefore, this commit changes
the test to first check if `/usr/bin/du` exists, and return early if it
does not, printing a warning indicating that the test is skipped.

If, someday, Rust's libtest grows a mechanism for tests to mark
themselves as skipped at runtime dynamically (rather than using the
`#[ignore]` attribute at compile-time), we should probably switch to
using that hypothetical mechanism here. See this
[internals.rust-lang.org thread][1].

If folks aren't happy with making the test skip itself, we *could*
change the implementation to worth with `du(1)` being at a different
location by finding it from the PATH, potentially falling back to
`/usr/bin/du` if it isn't on the PATH. My thought process was that it
was better to avoid complicating the production sled-agent code just to
support environments that the software will never actually run on for
test purposes, but I'm open to being convinced otherwise.

[1]: https://internals.rust-lang.org/t/pre-rfc-skippable-tests/14611

### CODE REVIEW

**Comment by hawkw** [comment]
Another option is changing the dev shell environment provided by the Nix flake to use [`buildFHSEnv`](https://ryantm.github.io/nixpkgs/builders/special/fhs-environments/) to create an isolated FHS root filesystem just for the Omicron dev shell...but that's annoying and I would prefer to avoid doing it if possible.

File: sled-agent/src/zone_bundle.rs [hunk 1]
```diff
@@ -1660,10 +1662,33 @@ async fn zfs_quota(path: &Utf8PathBuf) -> Result<u64, BundleError> {
...
+        //
+        // TODO(eliza): if, someday, Rust's libtest grows a mechanism for tests
+        // to mark themselves as skipped at runtime dynamically (rather than
+        // using the `#[ignore]` attribute at compile-time), we should use
+        // that here. See:
+        // https://internals.rust-lang.org/t/pre-rfc-skippable-tests/14611
+        if !Utf8Path::new(DU).exists() {
+            eprintln!("No {DU} executable found, skipping disk usage test!");
+            return;
+        }
```
Comment (bnaecker) [inline sled-agent/src/zone_bundle.rs:1690]: Could we put this behind a `#[cfg(target_os = "linux")]`? Super conservative, but would make me slightly more comfortable :)

**Review by bnaecker — APPROVED** [review]
With the caveat that I know zilch about NixOS, I approve!

File: sled-agent/src/zone_bundle.rs (hunk 1 above)
Comment (hawkw) [reply sled-agent/src/zone_bundle.rs:1690]: Sure, although, perhaps we want `#[cfg(not(target_os = "illumos"))]`? Would we also want this behavior on MacOS/Windows?

**Comment by hawkw** [comment]
Actually, here's a thought for another potential alternative: @sunshowers, is there a way to tell Nextest to skip a particular test using an environment variable? The Nix dev shell could set an env var that disables this test, so that it's only skipped on NixOS and you could get a message saying it was skipped, instead of a "PASS"...

**Comment by hawkw** [comment]
> Would we consider implementing the `du` behavior in Rust?

That would certainly be the ideal approach, especially given that there's already code for selecting different `du` args based on what OS we're on: https://github.com/oxidecomputer/omicron/blob/b78cbabd15f689ef92c87e0919dfc4c4dd82dc40/sled-agent/src/zone_bundle.rs#L1554-L1575

It would be a bit more work *just* to fix a test that doesn't work on my weird OS for insane people, but it also seems like a better solution in the long run; certainly less flaky than shelling out. On systems where there's a `du` binary, we could even cross-check our impl against `du(1)`...

File: sled-agent/src/zone_bundle.rs (hunk 1 above)
Comment (bnaecker) [reply sled-agent/src/zone_bundle.rs:1690]: I was just trying to preserve the existing behavior on the largest number of systems we already run the test on, that's all. Implementing `du` in Rust is also fine to me, just sounds like more work, but there may be a good library for it.

**Comment by hawkw** [comment]
I'm gonna take a crack at writing a pure-Rust `du(1)` and if I can get it working this morning, replace this with a PR for that instead.

**Comment by sunshowers** [comment]
> Actually, here's a thought for another potential alternative: @sunshowers, is there a way to tell Nextest to skip a particular test using an environment variable? The Nix dev shell could set an env var that disables this test, so that it's only skipped on NixOS and you could get a message saying it was skipped, instead of a "PASS"...

Long requested but not currently supported :(

**Comment by hawkw** [comment]
@davepacheco #6482 replaces the use of `du` with a Rust function. LMKWYT!

**Comment by hawkw** [comment]
> > Actually, here's a thought for another potential alternative: @sunshowers, is there a way to tell Nextest to skip a particular test using an environment variable? The Nix dev shell could set an env var that disables this test, so that it's only skipped on NixOS and you could get a message saying it was skipped, instead of a "PASS"...
> 
> Long requested but not currently supported :(

FWIW, I think this might be possible by adding a "nixos" profile to the nextest config file that skips the test, and having the devshell set `NEXTEST_PROFILE="nixos"`? But, I think Dave's approach of just RIIRing `du` is probably nicer anyway.

**Comment by sunshowers** [comment]
> > > Actually, here's a thought for another potential alternative: @sunshowers, is there a way to tell Nextest to skip a particular test using an environment variable? The Nix dev shell could set an env var that disables this test, so that it's only skipped on NixOS and you could get a message saying it was skipped, instead of a "PASS"...
> > 
> > Long requested but not currently supported :(
> 
> FWIW, I think this might be possible by adding a "nixos" profile to the nextest config file that skips the test, and having the devshell set `NEXTEST_PROFILE="nixos"`? But, I think Dave's approach of just RIIRing `du` is probably nicer anyway.

That would work but is kind of a big hammer, and the default set would need to stay up-to-date with this

**Pull request closed without merging**